addon/addon/models/animation-participant.ts

```diff
@@ -32,39 +32,129 @@ export class AnimationParticipantManager {
+                ...scope,
+                isOnDetachedBranch: true,
+              },
+              stop: false,
+            };
+          } else if (scope.isOnDetachedBranch) {
+            if (scope.lastLiveAncestor) {
+              nodesToGraft.set(node, scope.lastLiveAncestor);
+              return {
+                nextScope: {
+                  ...scope,
+                },
+                stop: true, // no need to iterate into the children of this node, because it is currently being used (it has animations on it)
+              };
+            } else {
+              if (node.parent) node.delete();
+              else this.DOMRefs = this.DOMRefs.filter((v) => v !== node);
+              deleted.push(node);
+              return {
+                nextScope: {
+                  ...scope,
+                },
+                stop: false,
+              };
+            }
+          } else {
+            // We're still on a live branch, so return this node as the lastLiveAncestor
+            // It will become the graft target later
+            return {
+              nextScope: {
+                ...scope,
+                lastLiveAncestor: node,
+              },
+              stop: false,
+            };
+          }
+        },
+        {
+          isOnDetachedBranch: false,
+          lastLiveAncestor: undefined,
+          stop: false,
+        }
+      );
+    }
+
+    for (let [nodeToGraft, graftTo] of nodesToGraft) {
+      nodeToGraft.parent = graftTo;
+      graftTo.children.push(nodeToGraft);
+    }
+
+    for (let node of deleted) {
+      // TODO: once we have cloning, we'll have to check the clones too
+      if (node.animationParticipant.uiState.detached?.DOMRef === node) {
+        node.animationParticipant.uiState.detached = undefined;
+      }
+    }
+
+    for (let animationParticipant of this.participants) {
       if (animationParticipant.canBeCleanedUp) {
-        assert(
-          'Animation participant to clean up has current uiState',
-          !animationParticipant.uiState.current
-        );
         this.participants.delete(animationParticipant);
-        animationParticipant.uiState.detached?.DOMRef?.delete();
-      } else {
-        if (
-          animationParticipant.uiState.detached &&
-          (!animationParticipant.uiState.detached.animation ||
-            animationParticipant.uiState.detached.animation.playState ===
-              'finished' ||
-            animationParticipant.uiState.detached.animation?.playState ===
-              'idle' ||
-            animationParticipant.uiState.detached.animation?.playState ===
-              'paused')
-        ) {
-          animationParticipant.uiState.detached?.DOMRef?.delete();
-          animationParticipant.uiState.detached = undefined;
-        }
-        // TODO: check that canceling the animations is happening at the right place
-        animationParticipant.cancelAnimations();
       }
+      animationParticipant.cancelAnimations();
```

Comment:
Cancel animations after we're done checking whether participants and DOMRefNodes need to be cleaned up.